### DISCUSSION TITLE
`RequiresTypeResolution` and `Configuration` annotations questions

### DISCUSSION THREAD

**Discussion by Whathecode** [detekt/detekt#4744] [Q&A] [answered]
The documentation of these annotations is limited. I have several questions:

1. Do the `RequiresTypeResolution` and `Configuration` annotations serve a purpose for custom rules, or only for the default detekt rules?
2. Are these annotations taken into account when placed on abstract class base rules?
3. If the [default recommended behavior for rules using type resolution is not to run](https://detekt.dev/type-resolution.html#writing-a-rule-that-uses-type-resolution), then what purpose does `RequiresTypeResolution` serve? Doesn't detekt simply ignore those rules already when type resolution is not enabled? It seems more correct behavior (and less boilerplate code) than to check for empty binding context in every rule requiring type resolution.

**Comment by marschwar** [comment, marked as answer]
>     1. Do the `RequiresTypeResolution` and `Configuration` annotations serve a purpose for custom rules, or only for the default detekt rules?

No, they do not. At least not at the moment. Both annotations are int the internal package and as such not meant to be used outside of detekt.

>     2. Are these annotations taken into account when placed on abstract class base rules?

I do not believe they are. Both are (at the moment) exclusively used auto generate the rule documentation.

>     3. If the [default recommended behavior for rules using type resolution is not to run](https://detekt.dev/type-resolution.html#writing-a-rule-that-uses-type-resolution), then what purpose does `RequiresTypeResolution` serve? Doesn't detekt simply ignore those rules already when type resolution is not enabled? It seems more correct behavior (and less boilerplate code) to check for empty binding context in every rule requiring type resolution.

You are right. Detekt could, but does not do so at the moment, exclude a rule that `RequiresTypeResolution` if there is no binding context. One reason for that is that there are rules that behave differently with and without type resolution enabled. That is an unpleasant issue we are aware of and also plan to address in the future.

**Reply by BraisGabin** [reply]
As @marschwar right now those annotations do nothing for custom rules. I would like to make them work on third party rules to autogenerate the config files now that they can be injected in the default configuration (documentation about this feature pending, it is on my TODO list)
